File: android/app/src/main/java/my/ssdid/wallet/feature/identity/IdentityDetailScreen.kt
package my.ssdid.wallet.feature.identity

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalClipboardManager
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.platform.LocalView
import my.ssdid.wallet.ui.components.HapticManager
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import my.ssdid.wallet.domain.SsdidClient
import my.ssdid.wallet.domain.model.Identity
import my.ssdid.wallet.domain.model.VerifiableCredential
import my.ssdid.wallet.domain.revocation.RevocationManager
import my.ssdid.wallet.domain.revocation.RevocationStatus
import my.ssdid.wallet.domain.vault.Vault
import my.ssdid.wallet.ui.theme.*
import java.time.Instant
import java.time.temporal.ChronoUnit
import javax.inject.Inject

@HiltViewModel
class IdentityDetailViewModel @Inject constructor(
    private val vault: Vault,
    private val ssdidClient: SsdidClient,
    private val revocationManager: RevocationManager,
    savedStateHandle: SavedStateHandle
) : ViewModel() {
    private val keyId: String = savedStateHandle["keyId"] ?: ""
    private val _identity = MutableStateFlow<Identity?>(null)
    val identity = _identity.asStateFlow()

    private val _credentials = MutableStateFlow<List<VerifiableCredential>>(emptyList())
    val credentials = _credentials.asStateFlow()

    private val _revocationResults = MutableStateFlow<Map<String, RevocationStatus>>(emptyMap())
    val revocationResults = _revocationResults.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error = _error.asStateFlow()

    private val _isDeactivating = MutableStateFlow(false)
    val isDeactivating = _isDeactivating.asStateFlow()

    init {
        viewModelScope.launch {
            val id = vault.getIdentity(keyId)
            _identity.value = id
            if (id != null) {
                val creds = vault.getCredentialsForDid(id.did)
                _credentials.value = creds
                val results = mutableMapOf<String, RevocationStatus>()
                for (vc in creds) {
                    results[vc.id] = revocationManager.checkRevocation(vc)
                }
                _revocationResults.value = results
            }
        }
    }

    fun deactivateIdentity(onComplete: () -> Unit) {
        viewModelScope.launch {
            _isDeactivating.value = true
            _error.value = null
            val result = ssdidClient.deactivateDid(keyId)
            _isDeactivating.value = false
            result.onSuccess {
                onComplete()
            }.onFailure { e ->
                io.sentry.Sentry.captureException(e)
                val body = if (e is retrofit2.HttpException) {
                    try { e.response()?.errorBody()?.string() } catch (_: Exception) { null }
                } else null
                _error.value = listOfNotNull(e.message, body).joinToString("\n")
                    .ifBlank { "Deactivation failed" }
            }
        }
    }

    fun clearError() { _error.value = null }
}

@Composable
fun IdentityDetailScreen(
    keyId: String,
    onBack: () -> Unit,
    onCredentialClick: (String) -> Unit,
    onRecoverySetup: (String) -> Unit = {},
    onKeyRotation: (String) -> Unit = {},
    onDeviceManagement: (String) -> Unit = {},
    viewModel: IdentityDetailViewModel = hiltViewModel()
) {
    val identity by viewModel.identity.collectAsState()
    val credentials by viewModel.credentials.collectAsState()
    val revocationResults by viewModel.revocationResults.collectAsState()
    val error by viewModel.error.collectAsState()
    val isDeactivating by viewModel.isDeactivating.collectAsState()
    val view = LocalView.current

    Column(Modifier.fillMaxSize().background(BgPrimary).statusBarsPadding().navigationBarsPadding()) {
        Row(
            Modifier.padding(start = 8.dp, end = 20.dp, top = 12.dp, bottom = 12.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            IconButton(
                onClick = onBack,
                modifier = Modifier.semantics { contentDescription = "Back" }
            ) { Text("\u2190", color = TextPrimary, fontSize = 20.sp) }
            Spacer(Modifier.width(4.dp))
            Text("Identity Details", style = MaterialTheme.typography.titleLarge)
        }

        var showDeactivateDialog by remember { mutableStateOf(false) }

        error?.let { msg ->
            Card(
                Modifier.fillMaxWidth().padding(horizontal = 20.dp, vertical = 4.dp),
                shape = RoundedCornerShape(8.dp),
                colors = CardDefaults.cardColors(containerColor = DangerDim)
            ) {
                Text(
                    msg,
                    modifier = Modifier.padding(12.dp),
                    color = Danger,
                    fontSize = 12.sp,
                    fontFamily = FontFamily.Monospace
                )
            }
        }

        identity?.let { id ->
            LazyColumn(
                Modifier.weight(1f).padding(horizontal = 20.dp),
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                item {
                    Card(
                        Modifier.fillMaxWidth(),
                        shape = RoundedCornerShape(16.dp),
                        colors = CardDefaults.cardColors(containerColor = BgCard)
                    ) {
                        Column(Modifier.padding(18.dp)) {
                            Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
                                Text(id.name, style = MaterialTheme.typography.headlineMedium)
                                Box(Modifier.clip(RoundedCornerShape(4.dp)).background(SuccessDim).padding(horizontal = 10.dp, vertical = 3.dp)) {
                                    Text("Active", fontSize = 11.sp, color = Success)
                                }
                            }
                            Spacer(Modifier.height(16.dp))

                            val clipboardManager = LocalClipboardManager.current
                            var didCopied by remember { mutableStateOf(false) }

                            // M6: Reset "Copied" label after 2 seconds
                            LaunchedEffect(didCopied) {
                                if (didCopied) {
                                    kotlinx.coroutines.delay(2000)
                                    didCopied = false
                                }
                            }

                            Column(Modifier.padding(vertical = 8.dp)) {
                                Row(
                                    Modifier.fillMaxWidth(),
                                    horizontalArrangement = Arrangement.SpaceBetween,
                                    verticalAlignment = Alignment.CenterVertically
                                ) {
                                    Text("DID", style = MaterialTheme.typography.labelSmall)
                                    TextButton(
                                        onClick = {
                                            clipboardManager.setText(AnnotatedString(id.did))
                                            didCopied = true
                                            HapticManager.success(view)
                                        },
                                        modifier = Modifier.semantics {
                                            contentDescription = if (didCopied) "DID copied" else "Copy DID"
                                        },
                                        contentPadding = PaddingValues(horizontal = 8.dp, vertical = 0.dp)
                                    ) {
                                        Text(
                                            if (didCopied) "Copied" else "Copy",
                                            fontSize = 12.sp,
                                            color = Accent
                                        )
                                    }
                                }
                                Text(
                                    id.did,
                                    fontSize = 13.sp,
                                    color = TextPrimary,
                                    fontFamily = FontFamily.Monospace
                                )
                                Spacer(Modifier.height(8.dp))
                                HorizontalDivider(color = Border)
                            }
                            DetailRow("Key ID", id.keyId, mono = true)
                            DetailRow("Algorithm", id.algorithm.name.replace("_", " "))
                            DetailRow("W3C Type", id.algorithm.w3cType, mono = true)
                            DetailRow("Created", id.createdAt)
                            DetailRow("Key Storage", "Hardware-backed")
                            DetailRow("Public Key", id.publicKeyMultibase.take(30) + "...", mono = true)
                        }
                    }
                }

                // Profile
                if (id.profileName != null || id.email != null) {
                    item {
                        Spacer(Modifier.height(4.dp))
                        Text("PROFILE", style = MaterialTheme.typography.labelMedium)
                        Card(
                            modifier = Modifier.fillMaxWidth(),
                            shape = RoundedCornerShape(12.dp),
                            colors = CardDefaults.cardColors(containerColor = BgCard)
                        ) {
                            Column(Modifier.padding(14.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
                                id.profileName?.let { name ->
                                    Row {
                                        Text("Name", fontSize = 14.sp, color = TextSecondary)
                                        Spacer(Modifier.weight(1f))
                                        Text(name, fontSize = 14.sp, color = TextPrimary)
                                    }
                                }
                                id.email?.let { email ->
                                    Row {
                                        Text("Email", fontSize = 14.sp, color = TextSecondary)
                                        Spacer(Modifier.weight(1f))
                                        Text(email, fontSize = 14.sp, color = TextPrimary)
                                    }
                                }
                            }
                        }
                    }
                }

                // Connected Services
                item {
                    Spacer(Modifier.height(8.dp))
                    Text("CONNECTED SERVICES", style = MaterialTheme.typography.labelMedium)
                    Spacer(Modifier.height(8.dp))
                }

                if (credentials.isEmpty()) {
                    item {
                        Card(
                            Modifier.fillMaxWidth(),
                            shape = RoundedCornerShape(12.dp),
                            colors = CardDefaults.cardColors(containerColor = BgCard)
                        ) {
                            Column(
                                Modifier.padding(24.dp).fillMaxWidth(),
                                horizontalAlignment = Alignment.CenterHorizontally
                            ) {
                                Text("\uD83D\uDD17", fontSize = 32.sp)
                                Spacer(Modifier.height(8.dp))
                                Text("No services connected", fontSize = 14.sp, color = TextSecondary)
                                Text("Scan a QR code to register with a service", fontSize = 12.sp, color = TextTertiary)
                            }
                        }
                    }
                } else {
                    items(credentials.size) { index ->
                        val vc = credentials[index]
                        val revocationStatus = revocationResults[vc.id]
                        val status = resolveDisplayStatus(vc, revocationStatus)
                        val name = serviceName(vc)
                        val url = serviceUrl(vc)
                        val statusColor = when (status) {
                            VcDisplayStatus.ACTIVE -> Success
                            VcDisplayStatus.EXPIRING -> Warning
                            VcDisplayStatus.EXPIRED -> Danger
                            VcDisplayStatus.REVOKED -> Danger
                        }
                        val statusLabel = when (status) {
                            VcDisplayStatus.ACTIVE -> "Active"
                            VcDisplayStatus.EXPIRING -> "Expiring soon"
                            VcDisplayStatus.EXPIRED -> "Expired"
                            VcDisplayStatus.REVOKED -> "Revoked"
                        }

                        Card(
                            Modifier.fillMaxWidth().clickable { onCredentialClick(vc.id) },
                            shape = RoundedCornerShape(12.dp),
                            colors = CardDefaults.cardColors(containerColor = BgCard)
                        ) {
                            Row(Modifier.padding(14.dp), verticalAlignment = Alignment.CenterVertically) {
                                Box(
                                    Modifier.size(10.dp).clip(CircleShape).background(statusColor)
                                )
                                Spacer(Modifier.width(12.dp))
                                Column(Modifier.weight(1f)) {
                                    Text(name, fontSize = 14.sp, fontWeight = FontWeight.Medium, color = TextPrimary, maxLines = 1)
                                    if (url != null) {
                                        Text(url, fontSize = 11.sp, color = TextTertiary, maxLines = 1)
                                    }
                                    Text("Issued: ${vc.issuanceDate.take(10)}", fontSize = 11.sp, color = TextTertiary)
                                }
                                Text(statusLabel, fontSize = 11.sp, color = statusColor)
                            }
                        }
                    }
                }

                // Action buttons
                item {
                    Spacer(Modifier.height(8.dp))
                    Text("ACTIONS", style = MaterialTheme.typography.labelMedium)
                    Spacer(Modifier.height(8.dp))
                }
                item {
                    Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                        ActionCard("\uD83D\uDD11", "Recovery", Modifier.weight(1f)) { onRecoverySetup(keyId) }
                        ActionCard("\uD83D\uDD04", "Rotate Key", Modifier.weight(1f)) { onKeyRotation(keyId) }
                        ActionCard("\uD83D\uDCF1", "Devices", Modifier.weight(1f)) { onDeviceManagement(keyId) }
                    }
                }

                // Danger zone
                item {
                    Spacer(Modifier.height(24.dp))
                    Text("DANGER ZONE", style = MaterialTheme.typography.labelMedium, color = Danger)
                    Spacer(Modifier.height(8.dp))

                    Button(
                        onClick = { showDeactivateDialog = true },
                        modifier = Modifier.fillMaxWidth().semantics {
                            contentDescription = if (isDeactivating) "Deactivating identity" else "Deactivate identity"
                        },
                        enabled = !isDeactivating,
                        colors = ButtonDefaults.buttonColors(containerColor = Danger),
                        shape = RoundedCornerShape(12.dp)
                    ) {
                        if (isDeactivating) {
                            CircularProgressIndicator(
                                modifier = Modifier.size(18.dp),
                                color = Color.White,
                                strokeWidth = 2.dp
                            )
                            Spacer(Modifier.width(8.dp))
                            Text("Deactivating...", color = Color.White)
                        } else {
                            Text("Deactivate Identity", color = Color.White)
                        }
                    }

                    Spacer(Modifier.height(24.dp))
                }
            }
        }

        if (showDeactivateDialog) {
            val serviceNames = credentials.map { serviceName(it) }
            val warningMessage = if (credentials.isEmpty()) {
                "This is irreversible. Your DID will be permanently " +
                    "deactivated on the registry and all associated " +
                    "data will be deleted."
            } else {
                "This identity is connected to ${credentials.size} " +
                    "service${if (credentials.size == 1) "" else "s"} " +
                    "(${serviceNames.joinToString(", ")}). " +
                    "You may lose access to your data and accounts on these services.\n\n" +
                    "This is irreversible. Your DID will be permanently deactivated on the registry."
            }
            AlertDialog(
                onDismissRequest = { showDeactivateDialog = false },
                title = { Text("Deactivate Identity?") },
                text = { Text(warningMessage) },
                confirmButton = {
                    Button(
                        onClick = {
                            showDeactivateDialog = false
                            viewModel.deactivateIdentity { onBack() }
                        },
                        colors = ButtonDefaults.buttonColors(containerColor = Danger)
                    ) {
                        Text("Deactivate", color = Color.White)
                    }
                },
                dismissButton = {
                    TextButton(onClick = { showDeactivateDialog = false }) {
                        Text("Cancel")
                    }
                }
            )
        }
    }
}

@Composable
fun ActionCard(icon: String, label: String, modifier: Modifier = Modifier, onClick: () -> Unit) {
    Card(
        modifier = modifier
            .clickable(onClick = onClick)
            .semantics { contentDescription = label },
        shape = RoundedCornerShape(12.dp),
        colors = CardDefaults.cardColors(containerColor = BgCard)
    ) {
        Column(
            Modifier.padding(18.dp).fillMaxWidth(),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Box(
                Modifier.size(42.dp).clip(RoundedCornerShape(12.dp)).background(AccentDim),
                contentAlignment = Alignment.Center
            ) {
                Text(icon, color = Accent, fontSize = 18.sp)
            }
            Spacer(Modifier.height(8.dp))
            Text(label, fontSize = 12.sp, fontWeight = FontWeight.Medium, color = TextPrimary)
        }
    }
}

enum class VcDisplayStatus { ACTIVE, EXPIRING, EXPIRED, REVOKED }

fun vcDisplayStatus(vc: VerifiableCredential): VcDisplayStatus {
    val exp = vc.expirationDate ?: return VcDisplayStatus.ACTIVE
    return try {
        val expInstant = Instant.parse(exp)
        val now = Instant.now()
        when {
            expInstant.isBefore(now) -> VcDisplayStatus.EXPIRED
            expInstant.isBefore(now.plus(30, ChronoUnit.DAYS)) -> VcDisplayStatus.EXPIRING
            else -> VcDisplayStatus.ACTIVE
        }
    } catch (_: Exception) {
        VcDisplayStatus.ACTIVE
    }
}

fun resolveDisplayStatus(vc: VerifiableCredential, revocationStatus: RevocationStatus?): VcDisplayStatus {
    if (revocationStatus == RevocationStatus.REVOKED) return VcDisplayStatus.REVOKED
    return vcDisplayStatus(vc)
}

fun serviceName(vc: VerifiableCredential): String {
    val props = vc.credentialSubject.additionalProperties
    // Prefer human-readable serviceName from VC (set by the service)
    val displayName = (props["serviceName"] as? kotlinx.serialization.json.JsonPrimitive)
        ?.takeIf { it.isString }?.content
    if (!displayName.isNullOrBlank()) return displayName
    // Fallback: service identifier, title-cased
    val serviceId = (props["service"] as? kotlinx.serialization.json.JsonPrimitive)
        ?.takeIf { it.isString }?.content
    if (!serviceId.isNullOrBlank()) {
        return serviceId.split(Regex("[ _-]"))
            .joinToString(" ") { it.replaceFirstChar { c -> c.uppercase() } }
    }
    // Last resort: truncate issuer DID
    val issuer = vc.issuer
    return if (issuer.length > 30) issuer.take(20) + "..." + issuer.takeLast(8) else issuer
}

fun serviceUrl(vc: VerifiableCredential): String? {
    return (vc.credentialSubject.additionalProperties["serviceUrl"] as? kotlinx.serialization.json.JsonPrimitive)
        ?.takeIf { it.isString }?.content?.ifBlank { null }
}

@Composable
fun DetailRow(label: String, value: String, mono: Boolean = false) {
    Column(Modifier.padding(vertical = 8.dp)) {
        Text(label.uppercase(), style = MaterialTheme.typography.labelSmall)
        Spacer(Modifier.height(2.dp))
        Text(
            value,
            fontSize = 13.sp,
            color = TextPrimary,
            fontFamily = if (mono) FontFamily.Monospace else FontFamily.Default
        )
        Spacer(Modifier.height(8.dp))
        HorizontalDivider(color = Border)
    }
}

File: ios/SsdidWallet/Feature/Identity/IdentityDetailScreen.swift
import SwiftUI

struct IdentityDetailScreen: View {
    @Environment(AppRouter.self) private var router
    @EnvironmentObject private var services: ServiceContainer

    let keyId: String

    @State private var identity: Identity?
    @State private var credentials: [VerifiableCredential] = []
    @State private var showDeactivateDialog = false
    @State private var isDeactivating = false
    @State private var didCopied = false
    @State private var errorMessage: String?

    var body: some View {
        VStack(spacing: 0) {
            // Header
            HStack(spacing: 4) {
                Button { router.pop() } label: {
                    Image(systemName: "chevron.left")
                        .foregroundStyle(Color.textPrimary)
                        .font(.system(size: 20))
                }
                .padding(.leading, 8)

                Text("Identity Details")
                    .font(.ssdidHeadline)
                    .foregroundStyle(Color.textPrimary)

                Spacer()
            }
            .padding(.vertical, 12)
            .padding(.trailing, 20)

            if let errorMessage {
                Text(errorMessage)
                    .font(.ssdidMono)
                    .foregroundStyle(Color.danger)
                    .padding(12)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .background(Color.dangerDim)
                    .cornerRadius(8)
                    .padding(.horizontal, 20)
                    .padding(.bottom, 4)
            }

            if let id = identity {
                ScrollView {
                    VStack(alignment: .leading, spacing: 8) {
                        // Main info card
                        VStack(alignment: .leading, spacing: 0) {
                            HStack {
                                Text(id.name)
                                    .font(.ssdidTitle)
                                    .foregroundStyle(Color.textPrimary)
                                Spacer()
                                Text("Active")
                                    .font(.system(size: 11))
                                    .foregroundStyle(Color.success)
                                    .padding(.horizontal, 10)
                                    .padding(.vertical, 3)
                                    .background(Color.successDim)
                                    .cornerRadius(4)
                            }

                            Spacer().frame(height: 16)

                            // DID with copy
                            HStack {
                                Text("DID")
                                    .font(.ssdidCaption)
                                    .foregroundStyle(Color.textTertiary)
                                Spacer()
                                Button {
                                    UIPasteboard.general.string = id.did
                                    HapticManager.notification(.success)
                                    didCopied = true
                                    Task {
                                        try? await Task.sleep(for: .seconds(2))
                                        didCopied = false
                                    }
                                } label: {
                                    Text(didCopied ? "Copied" : "Copy")
                                        .font(.ssdidCaption)
                                        .foregroundStyle(Color.ssdidAccent)
                                }
                            }
                            .padding(.vertical, 8)

                            Text(id.did)
                                .font(.system(size: 13, design: .monospaced))
                                .foregroundStyle(Color.textPrimary)

                            Spacer().frame(height: 8)
                            Divider().background(Color.ssdidBorder)

                            detailRow("Key ID", value: id.keyId, mono: true)
                            detailRow("Algorithm", value: id.algorithm.rawValue.replacingOccurrences(of: "_", with: " "))
                            detailRow("W3C Type", value: id.algorithm.w3cType, mono: true)
                            detailRow("Created", value: id.createdAt)
                            detailRow("Key Storage", value: "Hardware-backed")
                            detailRow("Public Key", value: String(id.publicKeyMultibase.prefix(30)) + "...", mono: true)
                        }
                        .ssdidCard()

                        // Profile
                        if id.profileName != nil || id.email != nil {
                            Spacer().frame(height: 4)
                            Text("PROFILE")
                                .font(.ssdidCaption)
                                .foregroundStyle(Color.textTertiary)

                            VStack(alignment: .leading, spacing: 8) {
                                if let profileName = id.profileName {
                                    HStack {
                                        Text("Name")
                                            .font(.system(size: 14))
                                            .foregroundStyle(Color.textSecondary)
                                        Spacer()
                                        Text(profileName)
                                            .font(.system(size: 14))
                                            .foregroundStyle(Color.textPrimary)
                                    }
                                }
                                if let email = id.email {
                                    HStack {
                                        Text("Email")
                                            .font(.system(size: 14))
                                            .foregroundStyle(Color.textSecondary)
                                        Spacer()
                                        Text(email)
                                            .font(.system(size: 14))
                                            .foregroundStyle(Color.textPrimary)
                                    }
                                }
                            }
                            .ssdidCard()
                        }

                        // Connected Services
                        Spacer().frame(height: 8)
                        Text("CONNECTED SERVICES")
                            .font(.ssdidCaption)
                            .foregroundStyle(Color.textSecondary)

                        if credentials.isEmpty {
                            VStack(spacing: 8) {
                                Text("🔗")
                                    .font(.system(size: 32))
                                Text("No services connected")
                                    .font(.system(size: 14))
                                    .foregroundStyle(Color.textSecondary)
                                Text("Scan a QR code to register with a service")
                                    .font(.system(size: 12))
                                    .foregroundStyle(Color.textTertiary)
                            }
                            .frame(maxWidth: .infinity)
                            .padding(24)
                            .background(Color.bgCard)
                            .cornerRadius(12)
                        } else {
                            ForEach(credentials, id: \.id) { vc in
                                let status = serviceConnectionStatus(vc)
                                let name = serviceName(vc)
                                let url = serviceUrl(vc)

                                HStack(spacing: 12) {
                                    Circle()
                                        .fill(status.color)
                                        .frame(width: 10, height: 10)
                                    VStack(alignment: .leading, spacing: 2) {
                                        Text(name)
                                            .font(.system(size: 14, weight: .medium))
                                            .foregroundStyle(Color.textPrimary)
                                        if let url = url {
                                            Text(url)
                                                .font(.system(size: 11))
                                                .foregroundStyle(Color.textTertiary)
                                                .lineLimit(1)
                                        }
                                        Text("Issued: \(String(vc.issuanceDate.prefix(10)))")
                                            .font(.system(size: 11))
                                            .foregroundStyle(Color.textTertiary)
                                    }
                                    Spacer()
                                    Text(status.label)
                                        .font(.system(size: 11))
                                        .foregroundStyle(status.color)
                                }
                                .padding(14)
                                .background(Color.bgCard)
                                .cornerRadius(12)
                            }
                        }

                        // Actions
                        Spacer().frame(height: 8)
                        Text("ACTIONS")
                            .font(.ssdidCaption)
                            .foregroundStyle(Color.textSecondary)

                        Spacer().frame(height: 8)

                        HStack(spacing: 10) {
                            actionCard(icon: "key.fill", label: "Recovery") {
                                router.push(.recoverySetup(keyId: keyId))
                            }
                            actionCard(icon: "arrow.triangle.2.circlepath", label: "Rotate Key") {
                                router.push(.keyRotation(keyId: keyId))
                            }
                            actionCard(icon: "iphone.gen2", label: "Devices") {
                                router.push(.deviceManagement(keyId: keyId))
                            }
                        }

                        // Danger zone
                        Spacer().frame(height: 24)
                        Text("DANGER ZONE")
                            .font(.ssdidCaption)
                            .foregroundStyle(Color.danger)

                        Spacer().frame(height: 8)

                        Button {
                            showDeactivateDialog = true
                        } label: {
                            if isDeactivating {
                                HStack(spacing: 8) {
                                    ProgressView()
                                        .progressViewStyle(CircularProgressViewStyle(tint: .white))
                                    Text("Deactivating...")
                                }
                            } else {
                                Text("Deactivate Identity")
                            }
                        }
                        .buttonStyle(.ssdidDanger)
                        .disabled(isDeactivating)

                        Spacer().frame(height: 24)
                    }
                    .padding(.horizontal, 20)
                }
            } else {
                Spacer()
                ProgressView()
                    .progressViewStyle(CircularProgressViewStyle(tint: .ssdidAccent))
                Spacer()
            }
        }
        .background(Color.bgPrimary)
        .alert("Deactivate Identity?", isPresented: $showDeactivateDialog) {
            Button("Cancel", role: .cancel) {}
            Button("Deactivate", role: .destructive) {
                deactivateIdentity()
            }
        } message: {
            if credentials.isEmpty {
                Text("This is irreversible. Your DID will be permanently deactivated on the registry and all associated data will be deleted.")
            } else {
                let serviceNames = credentials.map { serviceName($0) }
                let serviceList = serviceNames.joined(separator: ", ")
                Text("This identity is connected to \(credentials.count) service\(credentials.count == 1 ? "" : "s") (\(serviceList)). You may lose access to your data and accounts on these services.\n\nThis is irreversible. Your DID will be permanently deactivated on the registry.")
            }
        }
        .task {
            await loadIdentity()
            if let id = identity {
                credentials = await services.vault.getCredentialsForDid(id.did)
            }
        }
    }

    @ViewBuilder
    private func detailRow(_ label: String, value: String, mono: Bool = false) -> some View {
        VStack(alignment: .leading, spacing: 2) {
            Text(label.uppercased())
                .font(.ssdidCaption)
                .foregroundStyle(Color.textTertiary)
            Text(value)
                .font(mono ? .ssdidMono : .system(size: 13))
                .foregroundStyle(Color.textPrimary)

            Spacer().frame(height: 8)
            Divider().background(Color.ssdidBorder)
        }
        .padding(.vertical, 8)
    }

    @ViewBuilder
    private func actionCard(icon: String, label: String, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            VStack(spacing: 8) {
                ZStack {
                    RoundedRectangle(cornerRadius: 12)
                        .fill(Color.accentDim)
                        .frame(width: 42, height: 42)
                    Image(systemName: icon)
                        .font(.system(size: 18))
                        .foregroundStyle(Color.ssdidAccent)
                }
                Text(label)
                    .font(.system(size: 12, weight: .medium))
                    .foregroundStyle(Color.textPrimary)
            }
            .frame(maxWidth: .infinity)
            .padding(18)
            .background(Color.bgCard)
            .cornerRadius(12)
        }
    }

    private func loadIdentity() async {
        let vault = services.vault
        identity = await vault.getIdentity(keyId: keyId)
        if identity == nil {
            errorMessage = "Identity not found for key: \(keyId)"
        }
    }

    private enum ServiceConnectionStatus {
        case active, expiring, expired

        var color: Color {
            switch self {
            case .active: return .success
            case .expiring: return .warning
            case .expired: return .danger
            }
        }

        var label: String {
            switch self {
            case .active: return "Active"
            case .expiring: return "Expiring soon"
            case .expired: return "Expired"
            }
        }
    }

    private func serviceConnectionStatus(_ vc: VerifiableCredential) -> ServiceConnectionStatus {
        guard let exp = vc.expirationDate else { return .active }
        let formatter = ISO8601DateFormatter()
        guard let expDate = formatter.date(from: exp) else { return .active }
        let now = Date()
        if expDate < now { return .expired }
        if expDate < Calendar.current.date(byAdding: .day, value: 30, to: now) ?? now { return .expiring }
        return .active
    }

    private func serviceName(_ vc: VerifiableCredential) -> String {
        // additionalProperties values are AnyCodable — extract the underlying String
        if let anyCodable = vc.credentialSubject.additionalProperties["service"],
           let name = anyCodable.value as? String, !name.isEmpty {
            // Capitalize known service names for display
            switch name.lowercased() {
            case "drive": return "SSDID Drive"
            default: return name.capitalized
            }
        }
        // Fallback: truncate issuer DID
        let issuer = vc.issuer
        if issuer.count > 30 {
            return String(issuer.prefix(20)) + "..." + String(issuer.suffix(8))
        }
        return issuer
    }

    private func serviceUrl(_ vc: VerifiableCredential) -> String? {
        guard let anyCodable = vc.credentialSubject.additionalProperties["serviceUrl"],
              let url = anyCodable.value as? String, !url.isEmpty else { return nil }
        return url
    }

    private func deactivateIdentity() {
        isDeactivating = true
        Task {
            do {
                // Deactivate on registry + delete locally + log activity
                try await services.ssdidClient.deactivateDid(keyId: keyId)
                isDeactivating = false
                router.pop()
            } catch {
                isDeactivating = false
                errorMessage = error.localizedDescription
            }
        }
    }
}
